dace/libraries/mpi/nodes/redistribute.py

```diff
@@ -0,0 +1,115 @@
+# Copyright 2019-2022 ETH Zurich and the DaCe authors. All rights reserved.
+from dace import dtypes, library, properties, subsets, symbolic
+from dace.data import _prod
+from dace.libraries.mpi import utils
+from dace.sdfg import nodes
+from dace.transformation.transformation import ExpandTransformation
+from .. import environments
+from dace.codegen.targets import cpp
+from dace import subsets
+
+
+@library.expansion
+class ExpandRedistribute(ExpandTransformation):
+
+    environments = [environments.mpi.MPI]
+
+    @staticmethod
+    def expansion(node, parent_state, parent_sdfg):
+        inp_buffer, out_buffer = node.validate(parent_sdfg, parent_state)
+        redistr = parent_sdfg.rdistrarrays[node.redistr]
+        array_a = parent_sdfg.subarrays[redistr.array_a]
+        array_b = parent_sdfg.subarrays[redistr.array_b]
+
+        inp_symbols = [symbolic.symbol(f"__inp_s{i}") for i in range(len(inp_buffer.shape))]
+        out_symbols = [symbolic.symbol(f"__out_s{i}") for i in range(len(out_buffer.shape))]
+        inp_subset = subsets.Indices(inp_symbols)
+        out_subset = subsets.Indices(out_symbols)
+        inp_offset = cpp.cpp_offset_expr(inp_buffer, inp_subset)
+        out_offset = cpp.cpp_offset_expr(out_buffer, out_subset)
+        print(inp_offset)
+        print(out_offset)
+        inp_repl = ""
+        for i, s in enumerate(inp_symbols):
+            inp_repl += f"int {s} = __state->{node.redistr}_self_src[__idx * {len(inp_buffer.shape)} + {i}];\n"
+        out_repl = ""
+        for i, s in enumerate(out_symbols):
+            out_repl += f"int {s} = __state->{node.redistr}_self_dst[__idx * {len(out_buffer.shape)} + {i}];\n"
+        copy_args = ", ".join([
+            f"__state->{node.redistr}_self_size[__idx * {len(inp_buffer.shape)} + {i}], {istride}, {ostride}"
+            for i, (istride, ostride) in enumerate(zip(inp_buffer.strides, out_buffer.strides))
+        ])
+
+        code = f"""
+            int myrank;
+            MPI_Comm_rank(MPI_COMM_WORLD, &myrank);
+            MPI_Request* req = new MPI_Request[__state->{node._redistr}_sends];
+            MPI_Status* status = new MPI_Status[__state->{node._redistr}_sends];
+            MPI_Status recv_status;
+            if (__state->{array_a.pgrid}_valid) {{
+                for (auto __idx = 0; __idx < __state->{node._redistr}_sends; ++__idx) {{
+                    // printf("({redistr.array_a} -> {redistr.array_b}) I am rank %d and I send to %d\\n", myrank, __state->{node._redistr}_dst_ranks[__idx]);
```

Comment:
all these comments in the generated code are a bit cluttery (also appears in other expansions).

Reply:
Yes, it is, but they may be needed for quick debugging purposes the following weeks, so I left them there intentionally. They will be clean-up in a subsequent PR.